### PR TITLE
First steps towards implementing execution par_unseq

### PR DESCRIPTION
**Pull request by hkaiser**
This is related to #2271

### CODE REVIEW

File: hpx/parallel/util/loop.hpp [hunk 1]
```diff
@@ -83,21 +83,71 @@ namespace hpx { namespace parallel { namespace util
...
+            {
+#if defined(HPX_MSVC)
+#pragma loop(ivdep)
+#elif defined(HPX_GCC_VERSION)
+#pragma GCC ivdep
+#pragma omp simd
+#elif defined(HPX_INTEL_VERSION) || defined(HPX_CLANG_VERSION)
+#pragma ivdep
+#pragma omp simd
+#endif
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: I'd like to leave omp out of the picture here since we'd require users to compile with `-fopenmp` then.

Comment (sithhell) [reply hpx/parallel/util/loop.hpp]: What about other compilers, should we issue a warning?

**Comment by sithhell** [comment]
Would be nice to have some confidence if it really has the expected effects (better code gen) before pushing it further.

**Comment by hkaiser** [comment]
> Would be nice to have some confidence if it really has the expected effects (better code gen) before pushing it further.

I agree, but how can we get confidence without actually trying out things?

**Comment by sithhell** [comment]
Am 09.12.2017 10:30 nachm. schrieb "Hartmut Kaiser" <
notifications@github.com>:

Would be nice to have some confidence if it really has the expected effects
(better code gen) before pushing it further.

I agree, but how can we get confidence without actually trying out things?


Trying it out doesn't mean it has be merged ;)


—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/STEllAR-GROUP/hpx/pull/3063#issuecomment-350506901>,
or mute
the thread

**Comment by jeffhammond** [comment]
# Summary

You can't use OpenMP `simd` (or `for`) with a `!=` comparison for loop termination, i.e. the following is invalid (see below), at least as of OpenMP 4.5.
```c++
+#if defined(HPX_HAVE_OPENMP_SIMD)
 +#pragma omp simd
 +#endif
 +                for (/**/; it != end; ++it)
 +                {
 +                    f(it);
 +                }
```
OpenMP 5 is expected to fix this.  I don't know the precise status of that proposal but can look if you want.

I recall that Intel 18 compilers do not object to `!=` in an OpenMP `for` loop (haven't checked SIMD explicitly) but GCC 7 and Clang 5 do.

I would not expect a compiler to attempt to vectorize container accesses if a random access iterator is not supported...

# Details

The relevant portions of the [OpenMP 4.5 specification](http://www.openmp.org/wp-content/uploads/openmp-4.5.pdf) are:
* Section 2.8.1
> The simd directive places restrictions on the structure of the associated for-loops. Specifically, all associated for-loops must have canonical loop form (Section 2.6 on page 53).
* Section 2.6
> A loop has canonical loop form if it conforms to the following: 
> `for (init-expr; test-expr; incr-expr) structured-block`
> ...
> *test-expr* One of the following:
>                      *var relational-op b*
>                      *b relational-op var*
> ...
> *relational-op* One of the following:
>                      `<`
>                      `<=`
>                      `>`
>                      `>=`

**Comment by hkaiser** [comment]
@jeffhammond thanks for this information. This branch is really meant to be used for experimentation towards implementing `par_unseq`, and I admit that I don't have too much experience with this.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1390,6 +1390,16 @@ if(HPX_WITH_COMPILER_WARNINGS)
     hpx_add_compile_flag_if_available(-Werror=format)
     hpx_add_compile_flag_if_available(-Werror=missing-braces)
     hpx_add_compile_flag_if_available(-Werror=sign-compare)
+
+    # try to enable simd parallelization for par_unseq
+    find_package(OpenMP QUIET)
+    if(OPENMP_FOUND)
+      if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
+        hpx_add_compile_flag_if_available(-openmp-simd)
```
Comment (jeffhammond) [inline CMakeLists.txt]: `-openmp-simd` is deprecated (https://software.intel.com/en-us/node/693432).  It was replaced by `-qopenmp-simd` in version 16 or 17.

I recommend that CMake test for `-fopenmp-simd`, then `-qopenmp-simd` then `-openmp-simd`.  ICC supports many of the GCC flags for compatibility...

Comment (hkaiser) [reply CMakeLists.txt]: Ok, thanks.

File: hpx/parallel/util/prefetching.hpp [hunk 3]
```diff
@@ -253,9 +253,9 @@ namespace hpx { namespace parallel { namespace util
         HPX_FORCEINLINE void prefetch_addresses(T const& ... ts)
         {
             int const sequencer[] = {
-                (_mm_prefetch(
+                0, (_mm_prefetch(
```
Comment (jeffhammond) [inline hpx/parallel/util/prefetching.hpp]: For the non-x86 code path, you may want to use `__builtin_prefetch`, which both GCC and Clang support.

Comment (hkaiser) [reply hpx/parallel/util/prefetching.hpp]: We can do that, MSVC however supports `_mm_prefetch` only.

**Comment by jeffhammond** [comment]
@hkaiser These are important experiments, in part because the OpenMP standard working group recognizes that support for C++ in OpenMP is lacking.  While 5.0 is pretty close to finished, your experience in implementing PSTL will be useful in identifying gaps that should be addressed in the next iteration.

I don't know what code you allow yourself to look at (due to licensing), but both [RAJA](https://github.com/LLNL/RAJA) and [Intel PSTL](https://github.com/intel/parallelstl) may be useful.  In particular, RAJA supports a bunch of different pragmas for persuading compilers to vectorize inner loops.

**Comment by msimberg** [comment]
Closing this since it's not actively being worked on, but if someone feels like picking this up again, feel free to do so!

**Pull request closed without merging**